client/grpc/README.md

```diff
@@ -0,0 +1,74 @@
+# Kava gRPC Client
+
+The Kava gRPC client is a tool for making gRPC queries on a Kava chain.
+
+## Features
+
+- Easy-to-use gRPC client for the Kava chain.
+- Access all query clients for Cosmos and Kava modules using `client.Query` (e.g., `client.Query.Bank.Balance`).
+- Utilize utility functions for common queries (e.g., `client.BaseAccount(str)`).
+
+## Usage
+
```

Comment:
@DracoLi found out that trying to use this client from another package (kava's ethermint forked) failed with error during go dependency resolution,

```
⋊> ~/f/k/e/scratch on patch-ethermint-gas-tracking ⨯ go get github.com/kava-labs/kava/
client/grpc@260d43291e4321e1b6ddb18abc2381436c8f7bc5
go: downloading github.com/tendermint/tendermint v0.34.27
go: downloading github.com/gogo/protobuf v1.3.3
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/client/grpc/tmservice imports
        github.com/gogo/protobuf/gogoproto: reading github.com/gogo/protobuf/go.mod at revision v1.3.3: unknown revision v1.3.3
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/client/grpc/tmservice imports
        github.com/gogo/protobuf/grpc: reading github.com/gogo/protobuf/go.mod at revision v1.3.3: unknown revision v1.3.3
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/client/grpc/tmservice imports
        github.com/gogo/protobuf/proto: reading github.com/gogo/protobuf/go.mod at revision v1.3.3: unknown revision v1.3.3
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/client/grpc/tmservice imports
        github.com/gogo/protobuf/types: reading github.com/gogo/protobuf/go.mod at revision v1.3.3: unknown revision v1.3.3
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/client/grpc/tmservice imports
        github.com/tendermint/tendermint/abci/types: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/client/grpc/tmservice imports
        github.com/tendermint/tendermint/proto/tendermint/p2p: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/client/grpc/tmservice imports
        github.com/tendermint/tendermint/proto/tendermint/types: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/client/grpc/tmservice imports
        github.com/tendermint/tendermint/proto/tendermint/version: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/client/grpc/tmservice imports
        github.com/tendermint/tendermint/rpc/core/types: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/x/auth/types imports
        github.com/tendermint/tendermint/crypto: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/x/evidence/types imports
        github.com/tendermint/tendermint/crypto/tmhash: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/x/evidence/types imports
        github.com/tendermint/tendermint/libs/bytes: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/x/staking/types imports
        github.com/gogo/protobuf/protoc-gen-gogo/descriptor: reading github.com/gogo/protobuf/go.mod at revision v1.3.3: unknown revision v1.3.3
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/x/staking/types imports
        github.com/tendermint/tendermint/proto/tendermint/crypto: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/ibc-go/v6/modules/apps/transfer/types imports
        github.com/gogo/protobuf/jsonpb: reading github.com/gogo/protobuf/go.mod at revision v1.3.3: unknown revision v1.3.3
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/ibc-go/v6/modules/apps/transfer/types imports
        github.com/tendermint/tendermint/types: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/kava-labs/kava/x/bep3/types imports
        github.com/tendermint/tendermint/types/time: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/util imports
        github.com/kava-labs/kava/app imports
        github.com/tendermint/tendermint/libs/json: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/util imports
        github.com/kava-labs/kava/app imports
        github.com/tendermint/tendermint/libs/log: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/client/grpc/tmservice imports
        github.com/cosmos/cosmos-sdk/client imports
        github.com/tendermint/tendermint/libs/cli: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/client/grpc/tmservice imports
        github.com/cosmos/cosmos-sdk/client imports
        github.com/tendermint/tendermint/mempool: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/client/grpc/tmservice imports
        github.com/cosmos/cosmos-sdk/client imports
        github.com/tendermint/tendermint/rpc/client: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/client/grpc/tmservice imports
        github.com/cosmos/cosmos-sdk/client imports
        github.com/tendermint/tendermint/rpc/client/http: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/client/grpc/tmservice imports
        github.com/cosmos/cosmos-sdk/client/rpc imports
        github.com/tendermint/tendermint/p2p: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/x/auth/types imports
        github.com/cosmos/cosmos-sdk/crypto/codec imports
        github.com/tendermint/tendermint/crypto/encoding: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/x/auth/types imports
        github.com/cosmos/cosmos-sdk/crypto/codec imports
        github.com/tendermint/tendermint/crypto/sr25519: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/x/upgrade/types imports
        github.com/cosmos/cosmos-sdk/store/types imports
        github.com/tendermint/tendermint/crypto/merkle: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/util imports
        github.com/kava-labs/kava/app imports
        github.com/cosmos/cosmos-sdk/server/api imports
        github.com/tendermint/tendermint/rpc/jsonrpc/server: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/util imports
        github.com/kava-labs/kava/app imports
        github.com/cosmos/cosmos-sdk/x/genutil imports
        github.com/tendermint/tendermint/config: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/util imports
        github.com/kava-labs/kava/app imports
        github.com/cosmos/cosmos-sdk/x/genutil imports
        github.com/tendermint/tendermint/crypto/ed25519: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/util imports
        github.com/kava-labs/kava/app imports
        github.com/cosmos/cosmos-sdk/x/genutil imports
        github.com/tendermint/tendermint/libs/os: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/util imports
        github.com/kava-labs/kava/app imports
        github.com/cosmos/cosmos-sdk/x/genutil imports
        github.com/tendermint/tendermint/privval: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/util imports
        github.com/kava-labs/kava/app imports
        github.com/evmos/ethermint/server/config imports
        github.com/tendermint/tendermint/libs/strings: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/client/grpc/tmservice imports
        github.com/cosmos/cosmos-sdk/baseapp imports
        github.com/cosmos/cosmos-sdk/snapshots imports
        github.com/gogo/protobuf/io: reading github.com/gogo/protobuf/go.mod at revision v1.3.3: unknown revision v1.3.3
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/query imports
        github.com/cosmos/cosmos-sdk/x/upgrade/types imports
        github.com/cosmos/cosmos-sdk/store/types imports
        github.com/cosmos/cosmos-sdk/store/internal/proofs imports
        github.com/tendermint/tendermint/libs/rand: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/util imports
        github.com/kava-labs/kava/app imports
        github.com/cosmos/ibc-go/v6/modules/core imports
        github.com/cosmos/ibc-go/v6/modules/core/02-client/keeper imports
        github.com/tendermint/tendermint/light: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/util imports
        github.com/kava-labs/kava/app imports
        github.com/cosmos/ibc-go/v6/modules/core/02-client imports
        github.com/cosmos/ibc-go/v6/modules/light-clients/07-tendermint/types imports
        github.com/tendermint/tendermint/libs/math: reading github.com/tendermint/tendermint/go.mod at revision v0.34.27: unknown revision v0.34.27
go: github.com/kava-labs/kava/client/grpc imports
        github.com/kava-labs/kava/client/grpc/util imports
        github.com/kava-labs/kava/app imports
        github.com/kava-labs/kava/x/evmutil imports
        github.com/kava-labs/kava/x/evmutil/client/cli imports
        github.com/evmos/ethermint/crypto/hd imports
        github.com/btcsuite/btcutil/hdkeychain imports
        github.com/btcsuite/btcd/btcec: cannot find module providing package github.com/btcsuite/btcd/btcec
```

 only fix was to update the go.mod with the same replaces as we have in kava's go.mod

```
replace (
	// Use the cosmos keyring code
	github.com/99designs/keyring => github.com/cosmos/keyring v1.2.0
	// Use rocksdb 7.9.2
	github.com/cometbft/cometbft-db => github.com/kava-labs/cometbft-db v0.7.0-rocksdb-v7.9.2-kava.1
	// Use cosmos-sdk fork with backported fix for unsafe-reset-all, staking transfer events, and custom tally handler support
	github.com/cosmos/cosmos-sdk => github.com/kava-labs/cosmos-sdk v0.46.11-kava.1
	// See https://github.com/cosmos/cosmos-sdk/pull/13093
	github.com/dgrijalva/jwt-go => github.com/golang-jwt/jwt/v4 v4.4.2
	// Use ethermint fork that respects min-gas-price with NoBaseFee true and london enabled, and includes eip712 support
	github.com/evmos/ethermint => github.com/kava-labs/ethermint v0.21.1-0.20231127220940-f5eaf3a08c1a
	// See https://github.com/cosmos/cosmos-sdk/pull/10401, https://github.com/cosmos/cosmos-sdk/commit/0592ba6158cd0bf49d894be1cef4faeec59e8320
	github.com/gin-gonic/gin => github.com/gin-gonic/gin v1.7.0
	// Use the cosmos modified protobufs
	github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
	// Downgraded to avoid bugs in following commits which causes "version does not exist" errors
	github.com/syndtr/goleveldb => github.com/syndtr/goleveldb v1.0.1-0.20210819022825-2ae1ddf74ef7
	// Use cometbft fork of tendermint
	github.com/tendermint/tendermint => github.com/cometbft/cometbft v0.34.27
	// Indirect dependencies still use tendermint/tm-db
	github.com/tendermint/tm-db => github.com/kava-labs/tm-db v0.6.7-kava.3
)
```

probably worth calling out in the readme to save someone else time in resolving it (ideally this is why I would like to see this client in a seperate package, by keeping it in the kava monorepo we aren't really dogfooding the client and feeling / fixing pain points that a third party developer would go through cc @karzak @Behdad-Kava )

Reply:
also it not being a separate package is leading to import cycles trying to use this in ethermint as the client for talking to the cosmos sdk...ending up falling back to copy pasting all the code in this package for setting up the grpc connection, defeating the very purpose of this client in terms of being the one true / last implementation of a go client for talking to the kava grpc endpoints

Reply:
my two cents 😄 
attempting to use the grpc client in ethermint is not the same as including it in a service that interacts with the chain. i don't think a dependency of kava should attempt to import this grpc client. it's intended end user is a service interacting with the chain, not the chain interacting with itself

regarding the replaces, it's been my experience that you always need to replaces to match kava's. not sure if there's a more ergonomic way around it..